File: ga/gpt3_5_result_2024_6/result/server_clients_gpt/45_server_clients.c
use std::fs::{File, OpenOptions};
use std::io::prelude::*;
use std::os::unix::io::{AsRawFd, FromRawFd};
use std::ptr;
use std::ffi::CString;
use std::mem;
use std::thread;
use std::time::Duration;

fn main() {
    // Create and open a file
    let file = OpenOptions::new()
        .read(true)
        .write(true)
        .create(true)
        .open("temp")
        .unwrap();

    let fd = file.as_raw_fd();

    // Set the size of the file
    unsafe {
        libc::ftruncate(fd, 4096);
    }

    // Map the file into memory
    let ptr = unsafe {
        libc::mmap(ptr::null_mut(), 4096, libc::PROT_READ | libc::PROT_WRITE, libc::MAP_SHARED, fd, 0)
    };
    if ptr == libc::MAP_FAILED {
        panic!("mmap failed");
    }

    // Main loop
    let mut stop = 0;
    while stop < 10 {
        // Write to a specific offset within the memory-mapped area
        let p = (ptr as *mut u8).offset(1024);

        unsafe {
            libc::strcpy(p as *mut i8, CString::new("Test message").unwrap().as_ptr());
        }
        println!("Message written to the shared memory: {}", unsafe { CStr::from_ptr(p as *const i8).to_str().unwrap() });

        // Sleep to simulate delay and observe the output
        thread::sleep(Duration::from_secs(1));

        // Read from the same offset
        println!("Read from shared memory: {}", unsafe { CStr::from_ptr(ptr as *const i8).to_str().unwrap() });

        // Sleep again to simulate delay
        thread::sleep(Duration::from_secs(1));

        stop += 1;
    }

    // Unmap memory
    unsafe {
        libc::munmap(ptr, 4096);
    }

    // Close file and remove it
    drop(file); // File will be closed when out of scope

    match std::fs::remove_file("temp") {
        Ok(_) => {},
        Err(e) => panic!("Error removing file: {}", e),
    }
}

File: ga/4omini_result_2024_7/result/server_clients_mini/45_server_clients.rs
use std::fs::{OpenOptions, File};
use std::io::{self, Write};
use std::mem;
use std::os::unix::fs::FileExt;
use std::ptr;
use std::thread;
use std::time::Duration;
use memmap::{MmapMut, MmapOptions}; // Add this crate to Cargo.toml

fn main() -> io::Result<()> {
    // Create and open a file
    let file_path = "temp";
    let file = OpenOptions::new()
        .read(true)
        .write(true)
        .create(true)
        .open(file_path)?;

    // Set the size of the file
    file.set_len(4096)?;

    // Map the file into memory
    let mut mmap = unsafe { MmapOptions::new().len(4096).map_mut(&file)? };

    // Main loop
    for _ in 0..10 {
        // Write to a specific offset within the memory-mapped area
        let offset = 1024;
        let message = "Test message";
        
        // Ensure the string fits within the boundary
        let len = message.len().min(mmap.len() - offset - 1);
        
        unsafe {
            ptr::copy_nonoverlapping(message.as_ptr(), mmap.as_mut_ptr().add(offset), len);
            *mmap.as_mut_ptr().add(offset + len) = 0; // Null-terminate the string
        }

        println!("Message written to the shared memory: {}", unsafe {
            std::str::from_utf8_unchecked(&mmap[offset..offset + len])
        });

        thread::sleep(Duration::from_secs(1)); // Sleep to simulate delay and observe the output

        // Read from the same offset
        println!("Read from shared memory: {}", unsafe {
            std::str::from_utf8_unchecked(&mmap[offset..offset + len])
        });

        thread::sleep(Duration::from_secs(1)); // Sleep again to simulate delay
    }

    // Unmap is handled automatically when `mmap` goes out of scope
    drop(mmap);
    
    // Close the file by dropping it
    drop(file);

    // Unlink (delete) the file
    std::fs::remove_file(file_path)?;

    Ok(())
}